lib/react_client.dart

```diff
@@ -135,69 +162,371 @@ class ReactDartComponentFactoryProxy<TComponent extends Component> extends React
-      }
+    if (isProvider) {
+      propsForJs['value'] = ContextHelpers.jsifyNewContext(propsForJs['value']);
+    }
+
+    return propsForJs.jsObject;
+  }
+}
+
+/// Creates ReactJS [Function Component] from Dart Function.
+class ReactDartFunctionComponentFactoryProxy extends ReactComponentFactoryProxy with JsBackedMapComponentFactoryMixin {
+  /// The name of this function.
+  final String displayName;
+
+  /// The React JS component definition of this Function Component.
+  final JsFunctionComponent reactFunction;
+
+  ReactDartFunctionComponentFactoryProxy(this.reactFunction, [this.displayName]);
 
-      if (ref is Ref) {
-        propsForJs['ref'] = ref.jsRef;
+  @override
+  JsFunctionComponent get type => reactFunction;
+}
+
+/// Creates a function component from the given [dartFunctionComponent] that can be used with React.
+///
+/// [displayName] Sets the component name for debugging purposes.
+///
+/// In DDC, this will be the [DartFunctionComponent] name, but in dart2js it will be null unless
+/// overridden, since using runtimeType can lead to larger dart2js output.
+///
+/// This will result in the dart2js name being `ReactDartComponent2` (the
+/// name of the proxying JS component defined in _dart_helpers.js).
+JsFunctionComponent _wrapFunctionComponent(DartFunctionComponent dartFunctionComponent, {String displayName}) {
+  displayName ??= getProperty(dartFunctionComponent, 'name') ?? getProperty(dartFunctionComponent, '\$static_name');
+
+  // dart2js uses null and undefined interchangeably, meaning returning `null` from dart
+  // may show up in js as `undefined`, ReactJS doesnt like that and expects a js `null` to be returned,
+  // and throws if it gets `undefined`. `jsNull` is an interop variable that holds a JS `null` value
+  // to force `null` as the return value if user returns a Dart `null`.
+  // See: https://github.com/dart-lang/sdk/issues/27485
+  jsFunctionComponent(JsMap jsProps, [JsMap _legacyContext]) =>
+      dartFunctionComponent(JsBackedMap.backedBy(jsProps)) ?? jsNull;
+  JsFunctionComponent interopFunction = allowInterop(jsFunctionComponent);
+  if (displayName != null) {
+    // This is a work-around to display the correct name in the React DevTools.
+    _defineProperty(interopFunction, 'name', jsify({'value': displayName}));
+  }
+  // ignore: invalid_use_of_protected_member
+  setProperty(interopFunction, 'dartComponentVersion', ReactDartComponentVersion.component2);
+  return interopFunction;
+}
+
+/// Converts a list of variadic children arguments to children that should be passed to ReactJS.
+///
+/// Returns:
+///
+/// - `null` if there are no args and [shouldAlwaysBeList] is false
+/// - `[]` if there are no args and [shouldAlwaysBeList] is true
+/// - the single child if only one was specified
+/// - otherwise, the same list of args, will all top-level children validated
+dynamic _generateChildren(List childrenArgs, {bool shouldAlwaysBeList = false}) {
```

Comment:
This seems like it's just an improvement to the existing `_convertArgsToChildren`; could we combine the two implementations into one to reduce duplication of code? I'd vote for keeping the name `_convertArgsToChildren` since it's a little more descriptive.

Reply:
yeah so i originally replaced every FactoryProxy with my _generateChildren and _generateProps, and they worked for the most part (all tests in this repo pass) but in over_react there are like 2 tests that fail cause one expects a `empty` but gets `[null]` and i didnt want to deal with it so i removed it all thinking you probably wouldn't even like that much code changes lol

Reply:
I had originally set out in this PR to update all function factories to be able to use `_generateChildren` and `_generateJsProps` in order to reduce code duplication, it works, all the tests will pass but then in OverReact 2 or so tests fail because they expect `empty` but get `[null]` this was discovered on friday and rather than fix it i opted to revert it because of another issue with something else i previously did in this PR. so here we are. I could probably pursue it again and find the issue, it is probably something silly.